src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -66,68 +66,181 @@ public static int IndexOf(ref char searchSpace, int searchSpaceLength, ref char
         }
 
         [MethodImpl(MethodImplOptions.AggressiveOptimization)]
-        public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref char second, int secondLength)
+        public static unsafe int SequenceCompareTo(ref char firstStart, int firstLength, ref char secondStart, int secondLength)
         {
             Debug.Assert(firstLength >= 0);
             Debug.Assert(secondLength >= 0);
 
-            int lengthDelta = firstLength - secondLength;
-
-            if (Unsafe.AreSame(ref first, ref second))
+            if (Unsafe.AreSame(ref firstStart, ref secondStart))
                 goto Equal;
 
-            IntPtr minLength = (IntPtr)((firstLength < secondLength) ? firstLength : secondLength);
-            IntPtr i = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            int minLength = (firstLength < secondLength) ? firstLength : secondLength;
+            int offset = 0;
+
+            if (Sse2.IsSupported && minLength >= Vector128<ushort>.Count)
+            {
+                // >= Sse2 intrinsics are supported, and length is enough to use them so use that path.
+                // We jump forward to the intrinsics at the end of them method so a naive branch predict
+                // will choose the non-intrinsic path so short lengths which don't gain anything aren't
+                // overly disadvantaged by having to jump over a lot of code. Whereas the longer lengths
+                // more than make this back from the intrinsics.
+                goto IntrinsicsCompare;
+            }
 
-            if ((byte*)minLength >= (byte*)(sizeof(UIntPtr) / sizeof(char)))
+            // Only use the Vector path if intrinsics aren't supported.
+            if (!Sse2.IsSupported && Vector.IsHardwareAccelerated)
```

Comment:
I assume since these are all JIT intrinsics that there's no benefit to instead writing it as:
```C#
if (Sse2.IsSupported)
{
    if (minLength >= Vector128<ushort>.Count)
    {
        ...
    }
}
else
{
    if (Vector.IsHardwareAccelerated)
    {
        ...
    }
}
```
?

Reply:
Also, is it possible Sse2.IsSupported might return false but IsHardwareAccelerated might return true?  If yes, should we do the vector hardware accelerated test even if Sse2.IsSupported is false?

Reply:
> is it possible Sse2.IsSupported might return false but IsHardwareAccelerated might return true?

E.g. on ARM.

Reply:
I just read the code again and I'm not sure what I was asking: it's already doing the hardware accelerated check when IsSupported is false :)